File: pc-server/src/clipboard_monitor.py
"""Platform-specific clipboard monitoring."""
import platform
import time
import threading
from typing import Optional, Callable
import pyperclip


class ClipboardMonitor:
    """Monitors PC clipboard for changes."""
    
    def __init__(self, on_change: Callable[[str], None], debounce_ms: int = 500):
        """
        Initialize clipboard monitor.
        
        Args:
            on_change: Callback function called when clipboard changes (receives clipboard text)
            debounce_ms: Debounce time in milliseconds to avoid rapid-fire changes
        """
        self.on_change = on_change
        self.debounce_ms = debounce_ms / 1000.0  # Convert to seconds
        self._last_content: Optional[str] = None
        self._last_hash: Optional[str] = None
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._lock = threading.Lock()
        
    def start(self) -> None:
        """Start monitoring clipboard."""
        if self._running:
            return
            
        self._running = True
        self._thread = threading.Thread(target=self._monitor_loop, daemon=True)
        self._thread.start()
        print("Clipboard monitoring started")
        
    def stop(self) -> None:
        """Stop monitoring clipboard."""
        self._running = False
        if self._thread:
            self._thread.join(timeout=1.0)
        print("Clipboard monitoring stopped")
        
    def _monitor_loop(self) -> None:
        """Main monitoring loop."""
        while self._running:
            try:
                # Get current clipboard content
                current_content = pyperclip.paste()
                
                if current_content and current_content != self._last_content:
                    # Calculate hash to avoid duplicate notifications
                    import hashlib
                    current_hash = hashlib.sha256(current_content.encode('utf-8')).hexdigest()
                    
                    if current_hash != self._last_hash:
                        self._last_content = current_content
                        self._last_hash = current_hash
                        
                        # Debounce: wait a bit before notifying
                        time.sleep(self.debounce_ms)
                        
                        # Check if content still changed after debounce
                        if pyperclip.paste() == current_content:
                            self.on_change(current_content)
                            
            except Exception as e:
                print(f"Error monitoring clipboard: {e}")
                
            # Check every 100ms
            time.sleep(0.1)
            
    def get_current_content(self) -> Optional[str]:
        """Get current clipboard content."""
        try:
            return pyperclip.paste()
        except Exception:
            return None
            
    def set_content(self, content: str) -> None:
        """
        Set clipboard content (used to prevent loops).
        
        This updates internal state without triggering change callback.
        """
        with self._lock:
            self._last_content = content
            import hashlib
            self._last_hash = hashlib.sha256(content.encode('utf-8')).hexdigest()
            pyperclip.copy(content)

File: pc-server-dotnet/Clipboard/ClipboardMonitor.cs
using System.Threading;
using Microsoft.Extensions.Logging;
using AppConnectServer.Core;

namespace AppConnectServer.Clipboard;

public class ClipboardMonitor
{
    private readonly ILogger<ClipboardMonitor> _logger;
    private readonly Action<string> _onChange;
    private readonly int _debounceMs;
    private string? _lastContent;
    private string? _lastHash;
    private bool _running;
    private Thread? _monitorThread;
    private readonly object _lock = new();

    public ClipboardMonitor(
        ILoggerFactory loggerFactory,
        Action<string> onChange,
        int debounceMs = 500)
    {
        _logger = loggerFactory.CreateLogger<ClipboardMonitor>();
        _onChange = onChange;
        _debounceMs = debounceMs;
    }

    public void Start()
    {
        if (_running) return;

        _running = true;
        _monitorThread = new Thread(MonitorLoop)
        {
            IsBackground = true
        };
        _monitorThread.Start();
        _logger.LogInformation("Clipboard monitoring started");
    }

    public void Stop()
    {
        _running = false;
        _monitorThread?.Join(TimeSpan.FromSeconds(1));
        _logger.LogInformation("Clipboard monitoring stopped");
    }

    private void MonitorLoop()
    {
        while (_running)
        {
            try
            {
                var currentContent = ClipboardWriter.GetCurrentText();
                
                if (!string.IsNullOrEmpty(currentContent) && currentContent != _lastContent)
                {
                    var currentHash = EncryptionManager.CalculateHash(currentContent);
                    
                    if (currentHash != _lastHash)
                    {
                        lock (_lock)
                        {
                            _lastContent = currentContent;
                            _lastHash = currentHash;
                        }

                        // Debounce: wait a bit before notifying
                        Thread.Sleep(_debounceMs);

                        // Check if content still changed after debounce
                        if (ClipboardWriter.GetCurrentText() == currentContent)
                        {
                            _onChange(currentContent);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error monitoring clipboard");
            }

            Thread.Sleep(100); // Check every 100ms
        }
    }

    public string? GetCurrentContent()
    {
        return ClipboardWriter.GetCurrentText();
    }

    public void SetContent(string content)
    {
        lock (_lock)
        {
            _lastContent = content;
            _lastHash = EncryptionManager.CalculateHash(content);
            ClipboardWriter.WriteText(content);
        }
    }
}
